File: cc-1_wc-tool/main.py
# Implement the 'wc' unix tool which stands for 'word count'. It can count number of lines,
# words and characters and this is my Python implementation of the same
# This coding challenge can be found at: https://codingchallenges.fyi/challenges/challenge-wc

import os
import sys

# Helper functions

def printResults(results):
  for result in results:
    print(result)

# Step One: In this step your goal is to write a simple version of wc,
# let's call it ccwc (cc for Coding Challenges) that takes the command line option -c
# and outputs the number of bytes in a file.

def getBytes(paths: list) -> (list):
  results, totalBytes = [], 0
  for path in paths:
    try:
      byteSize = os.stat(path).st_size
    except FileNotFoundError:
      raise FileNotFoundError(f"Path {path} does not exist or is inaccessible")
    results.append(f"{byteSize} {path}")
    totalBytes += byteSize
  if len(paths) == 1:
    return results
  results.append(f"{totalBytes} total")
  return results

# < ---------------------------------------------------------------------------------------- >

# Step Two: In this step your goal is to support the command line option -l
# that outputs the number of lines in a file.

def getLines(paths: list) -> str:
  results, totalLines = [], 0
  for path in paths:
    try:
      fileHandle = open(path, 'r', encoding="utf8")
    except FileNotFoundError:
      raise FileNotFoundError(f"Path {path} does not exist or is inaccessible")
    lineCount = 0
    for _ in fileHandle:
      lineCount += 1
    results.append(f"{lineCount} {path}")
    totalLines += lineCount
  if len(paths) == 1:
    return results
  results.append(f"{totalLines} total")
  return results

# < ---------------------------------------------------------------------------------------- >

# Step Three: In this step your goal is to support the command line option -w
# that outputs the number of words in a file.

def getWords(paths: list) -> str:
  results, totalWords = [], 0
  for path in paths:
    try:
      fileHandle = open(path, 'r', encoding="utf8")
    except FileNotFoundError:
      raise FileNotFoundError(f"Path {path} does not exist or is inaccessible")
    wordCount = 0
    for line in fileHandle:
      wordCount += len(line.split())
    results.append(f"{wordCount} {path}")
    totalWords += wordCount
  if len(paths) == 1:
    return results
  results.append(f"{totalWords} total")
  return results

# < ---------------------------------------------------------------------------------------- >

# Step Four: In this step your goal is to support the command line option -m
# that outputs the number of characters in a file.

def getCharacters(paths: list) -> str:
  results, totalCharacters = [], 0
  for path in paths:
    try:
      fileHandle = open(path, 'r', encoding="utf8")
    except FileNotFoundError:
      raise FileNotFoundError(f"Path {path} does not exist or is inaccessible")
    charCount = len(fileHandle.read())
    results.append(f"{charCount} {path}")
    totalCharacters += charCount
  if len(paths) == 1:
    return results
  results.append(f"{totalCharacters} total")
  return results

# < ---------------------------------------------------------------------------------------- >

# Step Five: In this step your goal is to support the default option - 
# i.e. no options are provided, which is the equivalent to the -c, -l and -w options.

def getAll(paths: list) -> str:
  results = []
  totalCount = [0, 0, 0]
  for path in paths:
    linesResponse = int(getLines([path])[0].split()[0])
    wordsResponse = int(getWords([path])[0].split()[0])
    bytesResponse = int(getBytes([path])[0].split()[0])
    if os.path.exists(path) == False:
      raise FileNotFoundError(f"Path {path} does not exist or is inaccessible")
    results.append(f"{linesResponse} {wordsResponse} {bytesResponse} {path}")
    totalCount = [totalItemCount + itemCount for totalItemCount, itemCount in zip(totalCount, [linesResponse, wordsResponse, bytesResponse])]
  if len(paths) == 1:
    return results
  results.append(f"{totalCount[0]} {totalCount[1]} {totalCount[2]} total")
  return results
    

if __name__ == "__main__":

  cliInput = input('>')

  if cliInput.startswith('ccwc'):

    cliInput = cliInput.split()
    
    if len(cliInput) < 2:
      sys.exit("Not enough arguments passed")
    
    if cliInput[1] == '-c':
      printResults(getBytes(cliInput[2:]))
    elif cliInput[1] == '-l':
      printResults(getLines(cliInput[2:]))
    elif cliInput[1] == '-w':
      printResults(getWords(cliInput[2:]))
    elif cliInput[1] == '-m':
      printResults(getCharacters(cliInput[2:]))
    elif cliInput[1].startswith('-'):
      sys.exit("Incorrect option")
    else:
      printResults(getAll(cliInput[1:]))
  
  else:

    cliInput = cliInput.split('|')

    if len(cliInput) < 2:
      sys.exit("Not enough arguments passed")
    
    checkFilesPos = 1
    for i in range(1,len(cliInput)):
      if cliInput[i].strip().startswith('ccwc'):
        checkFilesPos = i - 1
        break
    
    files = cliInput[checkFilesPos].split()[1:]

    # Get option for ccwc
    cliInput = cliInput[checkFilesPos + 1].split()
    if len(cliInput) == 1:
      printResults(getAll(files))
    if cliInput[1] == '-c':
      printResults(getBytes(files))
    elif cliInput[1] == '-l':
      printResults(getLines(files))
    elif cliInput[1] == '-w':
      printResults(getWords(files))
    elif cliInput[1] == '-m':
      printResults(getCharacters(files))
    elif cliInput[1].startswith('-'):
      sys.exit("Incorrect option")

File: cc-1_wc-tool/main.go
// Implement the 'wc' unix tool which stands for 'word count'. It can count number of lines,
// words and characters and this is my Python implementation of the same
// This coding challenge can be found at: https://codingchallenges.fyi/challenges/challenge-wc

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Helper function

func printResults(getOutput []string) error {

	if len(getOutput) <= 0 {
		return fmt.Errorf("Error in printing results")
	}

	for _, output := range getOutput {
		fmt.Println(output)
	}

	return nil

}

func getBytes(filePaths []string) ([]string, error) {
	
	var results []string
	totalBytes := int64(0)

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer file.Close()

		// Get file statistics
		fileInfo, err := file.Stat()
		if err != nil {
			return nil, err
		}

		// Size of the file
		fileSize := fileInfo.Size()
		results = append(results, fmt.Sprintf("%d %s", fileSize, path))
		totalBytes += fileSize
	}

	if len(filePaths) == 1 {
		return results, nil
	}

	results = append(results, fmt.Sprintf("%d total", totalBytes))
	return results, nil

}

func getLines(filePaths []string) ([]string, error) {

	var results []string
	totalLines := 0

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer file.Close()

		scanner := bufio.NewScanner(file)
		lineCount := 0
		for scanner.Scan() {
			lineCount += 1
		}
		if err := scanner.Err(); err != nil {
			return nil, err
		}

		results = append(results, fmt.Sprintf("%d %s", lineCount, path))
		totalLines += lineCount
	}

	if len(filePaths) == 1 {
		return results, nil
	}

	results = append(results, fmt.Sprintf("%d total", totalLines))
	return results, nil

}

func getWords(filePaths []string) ([]string, error) {

	var results []string
	totalWords := 0

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer file.Close()

		scanner := bufio.NewScanner(file)
		wordCount := 0
		for scanner.Scan() {
			wordCount += len(strings.Fields(scanner.Text()))
		}
		if err = scanner.Err(); err != nil {
			return nil, err
		}

		results = append(results, fmt.Sprintf("%d %s", wordCount, path))
		totalWords += wordCount
	}

	if len(filePaths) == 1 {
		return results, nil
	}

	results = append(results, fmt.Sprintf("%d total", totalWords))
	return results, nil

}

func getCharacters(filePaths []string) ([]string, error) {

	var results []string
	totalCharacters := 0

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer file.Close()

		scanner := bufio.NewScanner(file)
		characterCount := 0
		for scanner.Scan() {
			characterCount += len(scanner.Text())
		}
		if err = scanner.Err(); err != nil {
			return nil, err
		}

		results = append(results, fmt.Sprintf("%d %s", characterCount, path))
		totalCharacters += characterCount
	}

	if len(filePaths) == 1 {
		return results, nil
	}

	results = append(results, fmt.Sprintf("%d total", totalCharacters))
	return results, nil

}

func getAll(filePaths []string) ([]string, error) {

	var results []string
	totalCount := []int{0, 0, 0}

	for _, path := range filePaths {
		file, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer file.Close()

		linesResult, err := getLines([]string{path})
		if err != nil {
			return nil, err
		}
		linesCount, err := strconv.Atoi(strings.Fields(linesResult[0])[0])
		if err != nil {
			return nil, err
		}

		wordsResult, err := getWords([]string{path})
		if err != nil {
			return nil, err
		}
		wordsCount, err := strconv.Atoi(strings.Fields(wordsResult[0])[0])
		if err != nil {
			return nil, err
		}

		bytesResponse, err := getBytes([]string{path})
		if err != nil {
			return nil, err
		}
		bytesCount, err := strconv.Atoi(strings.Fields(bytesResponse[0])[0])
		if err != nil {
			return nil, err
		}

		results = append(results, fmt.Sprintf("%d %d %d %s", linesCount, wordsCount, bytesCount, path))
		addTotal := []int{linesCount, wordsCount, bytesCount}
		for i := 0; i < 3; i++ {
			totalCount[i] += addTotal[i]
		}
	}

	if len(filePaths) == 1 {
		return results, nil
	}

	results = append(results, fmt.Sprintf("%d %d %d total", totalCount[0], totalCount[1], totalCount[2]))
	return results, nil

}

func getResults(option string, filePaths []string, defaultGetAllFlag bool, nonDirectCommand bool) error {

	if defaultGetAllFlag == true {
		getOutput, err := getAll(filePaths)
		if err != nil {
			return fmt.Errorf("Error getting all options %v", err)
		}

		if err = printResults(getOutput); err != nil {
			return fmt.Errorf("Error printing results %v", err)
		}

		return nil
	}

	if option == "-c" {
		if nonDirectCommand == false {
			filePaths = filePaths[1:]
		}
		getOutput, err := getBytes(filePaths)
		if err != nil {
			return fmt.Errorf("Error getting bytes in -c option %v", err)
		}

		if err = printResults(getOutput); err != nil {
			return fmt.Errorf("Error printing results %v", err)
		}
	} else if option == "-l" {
		if nonDirectCommand == false {
			filePaths = filePaths[1:]
		}
		getOutput, err := getLines(filePaths)
		if err != nil {
			return fmt.Errorf("Error getting lines in -l option %v", err)
		}
 
		if err = printResults(getOutput); err != nil {
			return fmt.Errorf("Error printing results %v", err)
		}
	} else if option == "-w" {
		if nonDirectCommand == false {
			filePaths = filePaths[1:]
		}
		getOutput, err := getWords(filePaths)
		if err != nil {
			return fmt.Errorf("Error getting words in -m option %v", err)
		}

		if err = printResults(getOutput); err != nil {
			return fmt.Errorf("Error printing results %v", err)
		}
	} else if option == "-m" {
		if nonDirectCommand == false {
			filePaths = filePaths[1:]
		}
		getOutput, err := getCharacters(filePaths)
		if err != nil {
			return fmt.Errorf("Error getting characters in -m option %v", err)
		}

		if err = printResults(getOutput); err != nil {
			return fmt.Errorf("Error printing results %v", err)
		}
	} else if strings.HasPrefix(option, "-") {
		return fmt.Errorf("Incorrect option")
	} else {
		getOutput, err := getAll(filePaths)
		if err != nil {
			return fmt.Errorf("Error getting all options %v", err)
		}

		if err = printResults(getOutput); err != nil {
			return fmt.Errorf("Error printing results %v", err)
		}
	}

	return nil
}

func main() {

	scanner := bufio.NewScanner(os.Stdin)

	// Take CLI input

	fmt.Printf("> ")

	scanner.Scan()
	cliInput := scanner.Text()

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading input: ", err)
		return
	}

	// Check if command starts with 'ccwc' or not
	if strings.HasPrefix(cliInput, "ccwc") {

		cliArgs := strings.Fields(cliInput)

		if len(cliArgs) < 2 {
			fmt.Println("Not enough arguments passed")
			return
		}

		option := cliArgs[1]
		if err := getResults(option, cliArgs[1:], false, false); err != nil {
			fmt.Println("Error getting results in ccwc command: ", err)
			return
		}

	} else {

		cliArgs := strings.Split(cliInput, "|")

		if len(cliArgs) < 2 {
			fmt.Println("Not enough arguments")
			return
		}

		checkFilePos := 1
		for i := 1; i < len(cliArgs); i++ {
			if strings.HasPrefix(strings.TrimSpace(cliArgs[i]), "ccwc") {
				checkFilePos = i - 1
				break
			}
		}

		filesPath := strings.Fields(cliArgs[checkFilePos])[1:]

		option := strings.Fields(cliArgs[checkFilePos + 1])

		if len(option) == 1 {
			if err := getResults(option[0], filesPath, true, true); err != nil {
				fmt.Println("Error in getting all options results: ", err)
				return
			}
		} else {
			if err := getResults(option[1], filesPath, false, true); err != nil {
				fmt.Println("Error in getting results: ", err)
				return
			}
		}
	}
}
